### PR TITLE
Typescript updates

### PR DESCRIPTION
**Pull request by jsnoble**
- converts to typescript
- id_reader now works off of a field

### CODE REVIEW

File: asset/src/elasticsearch_index_selector/processor.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+export default class IndexSelector extends BatchProcessor<IndexSelectorConfig> {
+    private formattedDate(record: DataEntity) {
+        const { date_field: dateField, timeseries } = this.opConfig;
+        let end = 10;
+
+        // date check
+        let date;
+
+        try {
+            date = new Date(record[dateField as string]).toISOString();
```
Comment (peterdemartini) [inline asset/src/elasticsearch_index_selector/processor.ts]: This doesn't validate the date correctly, use the [helper function](https://github.com/terascope/teraslice/blob/master/packages/utils/src/dates.ts#L14-L23) from `@terascope/utils`.

File: asset/src/elasticsearch_index_selector/processor.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+    private generateRequest(record: DataEntity, formatted: DataEntity[]) {
+        const indexSpec: IndexSpec = DataEntity.make({});
+        const index = this.indexName(record);
+        const meta: Partial<BulkMeta> = {
+            _index: index,
+            _type: this.opConfig.type
+        };
+
+        if (this.opConfig.preserve_id) meta._id = DataEntity.getMetadata(record, '_key');
```
Comment (peterdemartini) [inline asset/src/elasticsearch_index_selector/processor.ts]:
```suggestion
        if (this.opConfig.preserve_id) meta._id = record.getKey();
```

File: asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts [hunk 3]
```diff
@@ -0,0 +1,325 @@
+
+import { cloneDeep, toString } from '@terascope/job-components';
+import moment from 'moment';
+// @ts-ignore
+import parseError from '@terascope/error-parser';
```
Comment (peterdemartini) [inline asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts]: Use `parseError` from `@terascope/utils`.
```suggestion
import { parseError } from '@terascope/utils';
```

File: asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+                            list.push(cloneDeep(data));
+                            return iterate();
+                        }
+                        return resolve(list);
+                    })
+                    .catch((err) => {
+                        // retries happen at the idSlicer level
+                        const errMessage = parseError(err);
+                        logger.error('error trying to subslice by key on getIdData:', errMessage);
+                        reject(errMessage);
```
Comment (peterdemartini) [inline asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts]: Don't throw with a string and use TSError
```suggestion
                        reject(new TSError(err, {
                            reason: 'error trying to subslice by key'
                        }));
```

File: asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+
+        const { delayTime, interval } = slicerDates;
+        let startPoint = moment(slicerDates.start);
+        let limit = moment(slicerDates.end);
+        const dateArray: any[] = [];
+
+        logger.debug('all date configurations for date slicer', dateParams);
+
+        // set a timer to add the next set it should process
+        setInterval(() => {
```
Comment (peterdemartini) [inline asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts]: We need to clean up this interval on shutdown

File: asset/src/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
+// @ts-ignore
+import parseError from '@terascope/error-parser';
```
Comment (peterdemartini) [inline asset/src/helpers.ts]: Use `parseError` from `@terascope/utils`

```suggestion
import { parseError } from '@terascope/utils';
```

File: asset/src/id_reader/id-slicer.ts [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        }
+        return async function slicer() {
+            try {
+                const results = await determineKeySlice(gen, closePath, dateRange);
+                closePath = true;
+                return results;
+            } catch (err) {
+                const errMessage = parseError(err);
+                logger.error('id_slicer errored while making slice', errMessage);
+                const error = new Error(`Failure to make slice: ${toString(err)}`);
```
Comment (peterdemartini) [inline asset/src/id_reader/id-slicer.ts]: We should use `TSError` to wrap this error

File: test/elasticsearch_reader-spec.ts [hunk 8]
```diff
@@ -0,0 +1,791 @@
...
+                interval: '2hrs',
+                start: '2015-08-25T00:00:00',
+                end: '2015-08-25T00:02:00'
+            };
+            const errMsg = 'date_field_name: "date" for index: someindex does not exist';
+
+            try {
+                await makeSlicerTest(opConfig);
+            } catch (err) {
+                expect(startsWith(err.message, errMsg)).toEqual(true);
```
Comment (peterdemartini) [inline test/elasticsearch_reader-spec.ts]:
```suggestion
                expect(err.message).toStartWith(errMsg);
```

File: test/index_selector-spec.ts [hunk 9]
```diff
@@ -1,81 +1,110 @@
...
         };
 
-        const data = [
-            { _id: '1', date: '2019-07-02T00:00:00.001Z', name: 'bob' },
-        ];
+        const errMsg = 'elasticsearch_index_selector is mis-configured, if any of the following configurations are set: timeseries, index_prefix or date_field, they must all be used together, please set the missing parameters';
+        try {
+            await makeTest(op1);
+        } catch (err) {
+            expect(startsWith(err.message, errMsg)).toEqual(true);
```
Comment (peterdemartini) [inline test/index_selector-spec.ts]:
```suggestion
            expect(err.message).toStartWith(errMsg);
```

File: test/index_selector-spec.ts [hunk 10]
```diff
@@ -1,81 +1,110 @@
...
+        }
 
         try {
-            await opTest.processData(op1, data);
+            await makeTest(op2);
         } catch (err) {
-            expect(err.message).toBe('opConfig date field: name either does not exists or is not a valid date on the records processed');
+            expect(startsWith(err.message, errMsg)).toEqual(true);
```
Comment (peterdemartini) [inline test/index_selector-spec.ts]:
```suggestion
            expect(err.message).toStartWith(errMsg);
```

File: test/index_selector-spec.ts [hunk 11]
```diff
@@ -1,81 +1,110 @@
...
+        const opConfig = {
+            _op: 'elasticsearch_index_selector',
+            index: 'some-index'
+        };
+        const errMsg = 'type must be specified in elasticsearch index selector config if data is not a full response from elasticsearch';
+        try {
+            harness = WorkerTestHarness.testProcessor(opConfig, { assetDir, clients });
+            await harness.initialize();
+        } catch (err) {
+            expect(startsWith(err.message, errMsg)).toEqual(true);
```
Comment (peterdemartini) [inline test/index_selector-spec.ts]:
```suggestion
            expect(err.message).toStartWith(errMsg);
```

File: test/index_selector-spec.ts [hunk 12]
```diff
@@ -1,81 +1,110 @@
...
-        }).toThrowError('elasticsearch_index_selector is mis-configured, if any of the following configurations are set: timeseries, index_prefix or date_field, they must all be used together, please set the missing parameters');
-        expect(() => {
-            indexer.selfValidation(op2);
-        }).toThrowError('elasticsearch_index_selector is mis-configured, if any of the following configurations are set: timeseries, index_prefix or date_field, they must all be used together, please set the missing parameters');
-        expect(() => {
-            indexer.selfValidation(op3);
-        }).not.toThrowError();
-    });
+    it('will throw if other config options are not present with timeseries', async () => {
+        expect.hasAssertions();
```
Comment (peterdemartini) [inline test/index_selector-spec.ts]: This needs to be the number of assertions that expected to be done. It might be more use `await expect(makeTest(op1)).rejects.toThrowError(new RegExp(`^${errMsg}`));` for rejection tests
```suggestion
        expect.assertions(3);
```

File: test/simple_api_reader-spec.ts [hunk 13]
```diff
@@ -260,71 +307,162 @@ describe('simple_api_reader', () => {
...
-                    total: 2
-                });
+            function checkUpdate(updateObj: any) {
+                updatedConfig = get(updateObj, 'update[0]');
+                return true;
+            }
 
-            await harness.initialize([]);
+            beforeEach(async () => {
+                scope.get(/.*/)
```
Comment (peterdemartini) [inline test/simple_api_reader-spec.ts]: Why did you change this? Accepting any route can lead to tests passing but failing in real life.

```suggestion
                scope.get(`/${testIndex}`)
```

File: test/simple_api_reader-spec.ts [hunk 14]
```diff
@@ -260,71 +307,162 @@ describe('simple_api_reader', () => {
...
+                    .reply(200, {
+                        results: [{ created: start.toISOString() }],
+                        total: 1
+                    });
+
+                scope.get(/.*/)
```
Comment (peterdemartini) [inline test/simple_api_reader-spec.ts]: Why was this changed?

File: test/simple_api_reader-spec.ts [hunk 15]
```diff
@@ -260,71 +307,162 @@ describe('simple_api_reader', () => {
...
+                    .query(true)
+                    .reply(200, {
+                        results: [{ created: end.toISOString() }],
+                        total: 1
+                    });
+
+                scope.get(/.*/)
```
Comment (peterdemartini) [inline test/simple_api_reader-spec.ts]: Why was this changed?

File: test/mock_client.ts [hunk 16]
```diff
@@ -34,7 +52,8 @@ class MockClient {
         };
     }
 
-    async search() {
+    async search(query: AnyObject) {
+        this.searchQuery = query;
```
Comment (peterdemartini) [inline test/mock_client.ts]: Since we are mocking the elasticsearch client, ee should validate the search request here (like elasticsearch would).

File: test/mock_client.ts [hunk 17]
```diff
@@ -59,7 +78,5 @@ class MockClient {
     }
 
     // eslint-disable-next-line class-methods-use-this
-    async bulk(data) { return data; }
+    async bulk(data: any) { return data; }
```
Comment (peterdemartini) [inline test/mock_client.ts]: Since we are mocking the elasticsearch client, ee should validate the bulk request here (like elasticsearch would).

File: test/simple_api_reader-spec.ts (hunk 13 above)
Comment (jsnoble) [reply test/simple_api_reader-spec.ts]: Its not changed, this is just a preview error as the test was duplicated at that spot and the orignal one was moved, original test is kept as is with no changes

File: asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts [hunk 18]
```diff
@@ -0,0 +1,349 @@
...
+    ): Promise<SliceResults> {
+        const intervalNum = dateParams.interval[0];
+        const intervalUnit = dateParams.interval[1];
+
+        let count: number;
+        try {
+            count = await getCount(dateParams);
+        } catch (err) {
+            const errMessage = parseError(err);
+            logger.error('error with determine slice:', errMessage);
```
Comment (peterdemartini) [inline asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts]: This is error handling bad practice. (see main review comment)

File: asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts [hunk 19]
```diff
@@ -0,0 +1,349 @@
...
+                        if (data) {
+                            list.push(cloneDeep(data));
+                            return iterate();
+                        }
+                        return resolve(list);
+                    })
+                    .catch((err) => {
+                        // retries happen at the idSlicer level
+                        const errMessage = parseError(err);
+                        logger.error('error trying to subslice by key on getIdData:', errMessage);
```
Comment (peterdemartini) [inline asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts]: This is error handling bad practice. (see main review comment)

File: test/simple_api_reader-spec.ts (hunk 13 above)
Comment (peterdemartini) [reply test/simple_api_reader-spec.ts]: Regardless it is still too generic, it should be specific to the index. I would caution you against being lazy in tests, be lazy in the implementation, not your tests.

File: asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts [hunk 20]
```diff
@@ -0,0 +1,349 @@
...
+                dateParams.end = moment(data.end).add(intervalNum, intervalUnit);
+            }
+
+            if (shouldDivideByID && data.count >= threshold) {
+                logger.debug('date slicer is recursing by keylist');
+                return Promise.resolve(makeKeyList(data))
+                    .then((results) => results)
+                    .catch((err) => {
+                        const errMsg = parseError(err);
+                        logger.error('error while subslicing by key', errMsg);
```
Comment (peterdemartini) [inline asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts]: This is error handling bad practice. (see main review comment)

File: asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts [hunk 21]
```diff
@@ -0,0 +1,349 @@
...
+                dateParams.end = moment(data.end).add(step, unit);
+            }
+
+            if (shouldDivideByID && data.count >= threshold) {
+                logger.debug('date slicer is recursing by keylist');
+                return Promise.resolve(makeKeyList(data))
+                    .then((results) => results)
+                    .catch((err) => {
+                        const errMsg = parseError(err);
+                        logger.error('error while subslicing by key', errMsg);
```
Comment (peterdemartini) [inline asset/src/elasticsearch_reader/elasticsearch_date_range/slicer.ts]: This is error handling bad practice. (see main review comment)

File: asset/src/simple_api_reader/client.ts [hunk 22]
```diff
@@ -149,51 +149,49 @@ function createClient(context, opConfig) {
...
+                },
+                timed_out: false,
+                _shards: {
+                    total: 1,
+                    successful: 1,
+                    failed: 0
+                }
+            });
+        } catch (err) {
+            logger.error(`error while calling endpoint ${uri}, error: ${err.message}`);
```
Comment (peterdemartini) [inline asset/src/simple_api_reader/client.ts]: This is error handling bad practice. (see main review comment)
